docs: replace deprecated SwUpdate#available/activated with new API (#44858)
angular/angular: aio/content/guide/service-worker-communications.md

## `SwUpdate` service

The `SwUpdate` service gives you access to events that indicate when the service worker discovers an available update for your application.

The `SwUpdate` service supports three separate operations:

> * The *versionUpdates* to get notified about *detected*, *ready*, and *installation failed* events.

Author: I think *installed and read*` is not correct. The docs say:
... [VersionReadyEvent](https://angular.io/api/service-worker/VersionReadyEvent) event whenever a new version has been downloaded and is ready for activation.

```suggestion
* Getting notified when an updated version is *detected* on the server, *ready to install* or when an *installation fails*.
```

> * The *versionUpdates* to get notified about *detected*, *ready*, and *installation failed* events.

Reviewer: By "install" you seem to refer to the process of starting to serve a client with a specific version. I was referring to the fact that the necessary version assets have been cached by the SW and it is ready to be used to serve clients.

Both uses are a little arbitrary :grin:

In the context of SWs, the term "install" is referring to a [phase in the SW's lifecycle][1] after the SW script has been downloaded and before it is activated. The [install][2] event is usually used for downloading the necessary assets and populating the caches, before the SW is ready to serve clients.

Now, the Angular SW is a little different than usual SW, because the app version is decoupled from the SW script and each has their own lifecycle. So, when I refer to phases of app version lifecycle I try to keep as close as possible (conceptually) to SW script lifecycle.

Hence, I believe that "installing an app version" is the process of downloading the necessary assets and populating the caches to get the app version ready to be served to clients.

WDYT? (Do I even make sense? 😅)

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API#download_install_and_activate
[2]: https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/install_event

> * The *versionUpdates* to get notified about *detected*, *ready*, and *installation failed* events.

Author: Ok, I see what you mean. Indeed, the "normal" SW differs a little bit from an Angular SW. I now understand your suggestions and included them.

* Asking the service worker to check the server for new updates.
* Asking the service worker to activate the latest version of the application for the current tab.

### Version Updates

The `versionUpdates` omits three event types:

> * Getting notified of *detected* updates. These are new versions the service worker has detected and is about to start downloading it.
> * Getting notified of *ready* updates. These are new versions of the application to be loaded if the page is refreshed.
> * Getting notified of *installation failures*. This is when the service worker failed while installing a new version.

Author: ```suggestion
* `VersionDetectedEvent` is emitted when the service worker has detected a new version of the app on the server and is about to start downloading it.
* `VersionReadyEvent` is emitted when a new version of the app is available to be activated by clients.
* `VersionInstallationFailedEvent` is emitted when the installation of a new version failed.
  It may be used for logging/monitoring purposes.
```

> The `detected`, `ready` and `installation failed` events can be accessed via the `versionUpdates` `Observable` property of `SwUpdate`:

Author: ```suggestion
```

> <code-example path="service-worker-getting-started/src/app/log-update.service.ts" header="log-update.service.ts" region="sw-update"></code-example>

Reviewer: Could you update the example to show handling all event types? For example:

```ts
updates.versionUpdates.subscribe(evt => {
  switch (evt.type) {
    case 'VERSION_DETECTED':
      console.log(`Downloading new app version: ${evt.version.hash}`);
      break;
    case 'VERSION_READY':
      console.log(`New app version ready for use: ${evt.latestVersion.hash}`);
      break;
    case 'VERSION_INSTALLATION_FAILED':
      console.log(`Failed to install app version '${evt.version.hash}': ${evt.error}`);
      break;
  }
});
```

Author: Yes, I already thought about this. However, we have to decide first whether we want to stick to the `map` approach from the comment above or not to write an appropriate solution for all three cases

> Use the `versionUpdates` event to notify the user of a detected or pending update or to refresh their pages when the code they are running is out of date.
> Furthermore, you can subscribe to the *installation failed* event to log installation failures.

Author: ```suggestion
```

## `SwUpdate` service

The `SwUpdate` service gives you access to events that indicate when the service worker discovers and installs an available update for your application.

The `SwUpdate` service supports three separate operations:
* Getting notified when an updated version is *detected* on the server, *installed and ready* to be used locally or when an *installation fails*.
* Asking the service worker to check the server for new updates.
* Asking the service worker to activate the latest version of the application for the current tab.

### Version updates

The `versionUpdates` is an `Observable` property of `SwUpdate` and emits three event types:
* `VersionDetectedEvent` is emitted when the service worker has detected a new version of the app on the server and is about to start downloading it.

> * `VersionReadyEvent` is emitted when a new version of the app is available to be activated by clients.

Reviewer: It might be useful to include some suggestion on what this event can be used for. For example:

```suggestion
* `VersionReadyEvent` is emitted when a new version of the app is available to be activated by clients.
  It may be used to notify the user of an available update or prompt them to refresh the page.
```
